File: leetcode/easy/merge_two_sorted_list.go
package easy

import "github.com/johnazedo/leetcode/leetcode"

/*
Difficult: Easy
Link: https://leetcode.com/problems/merge-two-sorted-lists
*/
func mergeTwoLists(list1 *leetcode.ListNode, list2 *leetcode.ListNode) *leetcode.ListNode {
	head := &leetcode.ListNode{}
	var v1 int
	var v2 int
	temp := head
	for list1 != nil || list2 != nil {
		if list1 != nil {
			v1 = list1.Val
		} else {
			v1 = 999
		}
		if list2 != nil {
			v2 = list2.Val
		} else {
			v2 = 999
		}

		if v1 <= v2 {
			temp.Next = list1
			list1 = list1.Next
		} else {
			temp.Next = list2
			list2 = list2.Next
		}
		temp = temp.Next

	}
	return head.Next
}

File: leetcode/easy/merge_two_sorted_list.cpp
#include "../data_structures.h"
#include <bits/stdc++.h>

/*
 * Number: 21
 * Difficult: Easy
 * Link: https://leetcode.com/problems/merge-two-sorted-lists
 * Tags: Linked List, Recursion
 */ 

class RecursiveMergeTwoSortedList {
public:
    /*
    * Recursive approach
    *
    * Time: O(M+N)
    * Space: O(M+N) 
    * 
    * Time beats 85.5%
    * Memory beats 81.1%
    */
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if(list1 == NULL) return list2;
        if(list2 == NULL) return list1;
        
        if(list1->val <= list2->val) {
            list1->next = mergeTwoLists(list1->next, list2);
            return list1;
        } else {
            list2->next = mergeTwoLists(list1, list2->next);
            return list2;
        }
    }
};

// TODO: Make the iteratively approach
